library/Fission/CLI/Command/Down.hs

```diff
@@ -0,0 +1,72 @@
+-- | Grab files directly from IPFS
+module Fission.CLI.Command.Down (command, down) where
+
+import           RIO
+import           RIO.ByteString
+import           RIO.Directory
+import           RIO.Process (HasProcessContext)
+
+import           Data.Has
+import           Options.Applicative.Simple (addCommand)
+import           Options.Applicative (strArgument, metavar, help)
+import           System.Console.Haskeline
+import           System.Environment
+
+import           Fission.Internal.Constraint
+
+import qualified Fission.Storage.IPFS as IPFS
+import qualified Fission.IPFS.Types   as IPFS
+import qualified Fission.Web.Client   as Client
+
+import qualified Fission.CLI.Auth    as Auth
+import qualified Fission.CLI.Pin     as CLI.Pin
+import           Fission.CLI.Config.Types
+import qualified Fission.CLI.Display.Cursor  as Cursor
+import qualified Fission.CLI.Display.Success as CLI.Success
+import qualified Fission.CLI.Display.Error   as CLI.Error
+import qualified Fission.CLI.Display.Wait    as CLI.Wait
+
+-- | The command to attach to the CLI tree
+command :: MonadUnliftIO m
+        => HasLogFunc        cfg
+        => HasProcessContext cfg
+        => Has IPFS.BinPath  cfg
+        => Has IPFS.Timeout  cfg
+        => Has Client.Runner cfg
+        => cfg
+        -> CommandM (m ())
+command cfg =
+  addCommand
+    "down"
+    "pull a ipfs or ipns object down to your system"
+    (\cid -> runRIO cfg $ down cid)
+    (strArgument (
+      metavar "ContentID" <> help "The CID of the IPFS object you want to download"
+    ))
+
+-- | Sync the current working directory to the server over IPFS
+down :: MonadRIO        cfg m
+   => HasLogFunc        cfg
+   => HasProcessContext cfg
+   => Has IPFS.Timeout  cfg
+   => Has IPFS.BinPath  cfg
+   => Has Client.Runner cfg
+   => IPFS.CID
+   -> m ()
+down cid = do
+  cfg <- ask
+  getResult <- liftIO
+              . CLI.Wait.waitFor "Retrieving Object..."
+              . runRIO cfg
+              $ IPFS.getContent cid
+
+  case getResult of
+    Right _ok -> do
+      let
+        rawCID = IPFS.unaddress cid
+        successMessage = rawCID <> " Successfully downloaded!"
+
+      CLI.Success.putOk successMessage
+
+    Left  err ->
+      CLI.Error.put err "Oh no! The download failed unexpectedly"
```

Comment:
This feels a bit wordy. Have you considered:

```haskell
--   Destructuring
--   vvvvvvvvvvvvvv
down cid@(CID hash) = do
  cfg <- ask
  getResult <- liftIO
              . CLI.Wait.waitFor "Retrieving Object..."
              . runRIO cfg
              $ IPFS.getContent cid

  case getResult of --   Use it here vvvv
    Right _ok -> CLI.Success.putOk $ hash <> " Successfully downloaded!"
    Left  err -> CLI.Error.put err "Oh no! The download failed unexpectedly"
```

or even

```haskell
down cid@(CID hash) = do
  cfg <- ask
  liftIO (CLI.Wait.waitFor "Retrieving Object..." $ runRIO cfg $ IPFS.getContent cid) >>= \case
    Right _ok -> CLI.Success.putOk $ hash <> " Successfully downloaded!"
    Left  err -> CLI.Error.put err "Oh no! The download failed unexpectedly"
```

I'm going to get rid of that `liftIO` in a quick PR in a second here...

Reply:
I'm tempted to go with
```haskell
down cid@(IPFS.CID hash) = cid
  & IPFS.getContent
  & CLI.Wait.waitFor "Retrieving Object..."
  >>= \case
    Right _ok ->
      CLI.Success.putOk $ hash <> " Successfully downloaded!"
    Left  err ->
      CLI.Error.put err "Oh no! The download failed unexpectedly"
```
😛
however I think this is better
```haskell
down cid@(IPFS.CID hash) = do
  getResult <- CLI.Wait.waitFor "Retrieving Object..."
              $ IPFS.getContent cid

  case getResult of
    Right _ok ->
      CLI.Success.putOk $ hash <> " Successfully downloaded!"
    Left  err ->
      CLI.Error.put err "Oh no! The download failed unexpectedly"
```
as it clearly separates into "during computation" and after

Reply:
actually... let me know your thoughts on my obvious elixir style!

Reply:
@bechurch yeah, it's actually harder to read IMO. You're working in different levels, it's noisier, and so on. I totally get that you're used to thinking in pipes, but you won't be able to rely on this style in the long term — there's lots of code that will have a higher cognitive overhead by trying to pipe like this.

I've been thinking about this a bit. If we as a team decide to go totally off pf the idiomatic style and use pipes, we should consider the Elm style. It's less elegant and noisier, but is I think the most readable pipe style for this language family:

```haskell
down cid@(IPFS.CID hash) = 
  cid
    |> IPFS.getContent
    |> CLI.Wait.waitFor "Retrieving Object..."
    |> bind \case
      Right _ok ->
        CLI.Success.putOk $ hash <> " Successfully downloaded!"
      Left  err ->
        CLI.Error.put err "Oh no! The download failed unexpectedly"
```